cmd/cloudFoundryDeleteService.go

```diff
@@ -30,10 +37,55 @@ func cloudFoundryDeleteService(CloudFoundryDeleteServiceOptions cloudFoundryDele
 	return nil
 }
 
-func cloudFoundryLogin(CloudFoundryDeleteServiceOptions cloudFoundryDeleteServiceOptions, c execRunner) error {
-	var cfLoginScript = []string{"login", "-a", CloudFoundryDeleteServiceOptions.CfAPIEndpoint, "-o", CloudFoundryDeleteServiceOptions.CfOrg, "-s", CloudFoundryDeleteServiceOptions.CfSpace, "-u", CloudFoundryDeleteServiceOptions.Username, "-p", CloudFoundryDeleteServiceOptions.Password}
+func cloudFoundryDeleteServiceKeys(options cloudFoundryDeleteServiceOptions, c execRunner) error {
+
+	var cfFindServiceKeysScript = []string{"service-keys", options.CfServiceInstance}
+
+	var serviceKeyBytes bytes.Buffer
+	c.Stdout(&serviceKeyBytes)
+
+	err := c.RunExecutable("cf", cfFindServiceKeysScript...)
+	if err != nil {
+		cloudFoundryLogout(c)
+		log.Entry().
+			WithError(err).
+			Fatal("Failed to Delete Service Key, most likely your service doesn't exist")
```

Comment:
After more consideration, I believe you should not exit via the logger in this function. It already returns an error, and I think ignoring an error returned from a function is at least a warning in Go. Best would be to have the cloudFoundryLogout() call only once -- and in the function where you call cloudFoundryLogin() for symmetry. The coding pattern which I have adopted (but I am not sure whether that is ideal), is to declare the error first, then code the happy-path inside if-blocks, like so:
```
var err error
if err == nil {
    err = someFunction()
}
if err == nil {
    err = someOtherFunction()
}
...
if err != nil {
    cloudFoundryLogout()
    log.Entry().WithError(err).Fatal(...)
}
```
But in fact, you should not ignore the error from cloudFoundryLogin() in the first place, so you could write:
```
err := cloundFoundryLogin()
if err != nil {
    log.... Fatal()
}
```
You just need to be careful with assignment `=` versus declaration `:=` especially if you need to assign err inside an if-block. If you use the short declaration+assignment, you actually shadow "err" inside the if-block and the outer err remains nil.

Reply:
(Of course it would be even nicer if we could defer cloudFoundryLogout() directly after logging in successfully, but this clashes with exiting without running defer`d functions.)

Reply:
pattern which we applied so far:
- bubble up errors until the entry function, in this case `cloudFoundryDeleteService()`
- if error then use `log.Entry().Fatal()`

We should follow this here as well

Reply:
Then you could use `defer` and run the logout in a closure.

Reply:
I don't quite get the benefit of your recommendation. The error from cloudFoundryLogin() is currently already handled in the following implementation:
```
if err != nil {
	log.Entry().
		WithError(err).
		Fatal("Failed to login to Cloud Foundry")
}
log.Entry().Info("Logged in successfully to Cloud Foundry..")
return err
```
With the current implementation it's possible to log where error occured and what the error is. Your recommendation would only allow one generic exit for all errors. True, it would throw the correct error but no information regarding which step went wrong.
Where is the difference between bubbling up the errors or throwing them directly where they occur?

Reply:
One problem is that when exiting via the logging framework, deferred functions are not run. That is why you need to take care of logging out of cf in several places before logging with .Fatal(). This is fragile and violates the principle that a function should only have one concern. Logging out should happen at the level where logging in happened.
The code is just not as nice as it could be if you have a function that returns an error, but that error is actually ignored, and instead fatal errors exit.
If you care concerned about bubbling up more specific errors, you can use error wrapping (google it). One easy way to use wrapping while bubbling up is to `return fmt.Errorf("error occured: %w", err)` The "%w" triggers the generation of an error with its own message that wraps "err".

Reply:
Sorry, I didn't answer more directly to the code you pointed out: Why do you declare `cloudFoundryLogin()` as returning `error` if it never returns anything but `nil`? I would prefer that you keep it returning an error, but actually do so, and then bubble up the error as Oliver recommended. This makes the code more robust for future changes when someone takes the return values from functions as guidance for their expected behavior.

Reply:
Just to add some context:

We made the design decision to use `Fatal()` of the logging framework to exit the step.
This is accompanied with respective handling of post actions once such a fatal error occurs.

This implies that the step cleans up its own stuff before calling `Fatal()`.
Ideally this is done centrally in the "entry" function so that e.g. a cf logout can happen also centrally.

If the concern is about the location of error the design decision here is to use the package `github.com/pkg/errors` with the function `errors.Wrap()`.
This will create transparency in where the error occured.

Reply:
Ok got it now, instead of logging out separately you would just catch/bubble up the errors and if none exist you simply log out. If errors occur you log out + log the error message & exit with fatal.
Thank you for the hint :-)

I changed the code according to your recommendations to bubble up the errors in the separate functions and to handle them "centrally" where logging in happens. Can you take a look at the code again if this matches your/projects expectations?